[Core] Support image processor

DarkLight1337 opened the pull request:
I have implemented a plugin architecture (`MultiModelPlugin`) over `MultiModalData` to define how each modality type should be preprocessed before being passed to the model as keyword arguments. This preserves the contract between the output of HuggingFace processor and the input into the HuggingFace model. As long as those keyword arguments do not conflict with the ones we have in vLLM, I think this is a good way to make the framework flexible enough to support other multi-modal architectures.

FIX #4054 (the data is now automatically converted into the model's device)

### Related Contributions

This PR is part of #3978.

This PR also implements Proposals 1 and 3 of #4194.

### Features
1. Refactor of `MultiModalData`
    - Image input is now split into two subclasses:
        - `ImageFeatureData` represents the image features of LLaVA after being passed through the vision tower, but before the multi-modal projection is applied.
        - `ImagePixelData` represents the raw image (using `PIL.Image` class). `AutoImageProcessor` from HuggingFace is loaded from `config.json` to pre-process input images before being passed to the model as `pixel_values`. As with the tokenizer, you can override the default one and specify the version of image processor via `EngineConfig`; you can even disable image preprocessing altogether, which is useful if you want to pass in images that have already been preprocessed.
    - The LLaVA implementation has been updated accordingly to accept the new inputs.
2. A new documentation page for using VLMs can be found under ([dev/multimodal](https://github.com/DarkLight1337/vllm-rocm/blob/mm-data-processor/docs/source/dev/multimodal.rst)).

### Compatibility Changes

- `pillow` will be upgraded to a `common` dependency (from `dev`) to process the images.

DarkLight1337 committed:
Support image processor
- Also add docs for basic VLM usage

DarkLight1337 committed:
Convert dtype in multi modal processing
- Other data types may need to be of different dtype from that of the model

DarkLight1337 commented:
The LLaVA test passes on my end (with the output matching the HF output shown in CI). Does anyone have a clue what might cause it to fail in CI? Perhaps a case of floating-point error in GPU computation?

DarkLight1337 committed:
Move `MultiModalData` to new subpackage `multimodal`

DarkLight1337 committed:
Add multi-modal processor registry

DarkLight1337 committed:
Initialize the processor only once

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Move processor to model runner

DarkLight1337 committed:
Refactor registry to plugin pattern in order to support specifying dummy data

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Combine prompt inputs

DarkLight1337 committed:
Fix a bunch of tests

DarkLight1337 committed:
Fix LLaVA test

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Fix `benchmark_latency` test

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Clarify tokenizer usage

DarkLight1337 committed:
Rename `encode_request -> process_model_inputs`

DarkLight1337 committed:
Support old API in `LLM.generate`

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Fix import error

DarkLight1337 committed:
Add tests to ensure old API still works
- To facilitate equality tests, `CompletionOutput` is now a dataclass

DarkLight1337 committed:
Let all entrypoints tests be run at the same time

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Add tests for LLM.encode and fix corresponding bugs

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Rename `_add_requests` to `_validate_and_add_requests` to be more similar to the original `_validate_and_prepare_requests`

DarkLight1337 committed:
Separate `entrypoints` tests into two groups

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Remove duplicate comment

DarkLight1337 committed:
Fix memory profiling error

DarkLight1337 committed:
Fix memory usage for embedding server

DarkLight1337 committed:
Update embeddings API to use new imputs

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Simplify code and fix type annotations

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Merge branch 'upstream' into llm-inputs

DarkLight1337 committed:
Avoid duplicate `Tensor.to` calls

DarkLight1337 committed:
Merge `llm` groups back into one by enabling gc

DarkLight1337 committed:
Add test for image pixel processor

ywang96 reviewed (commented):
Thank you for the great work and I left a few comments! I've also reviewed your other PR https://github.com/vllm-project/vllm/pull/4910, so let's get that merged first then come back to this.

P.S I'm also working on a GPT4V API server PR as a follow up once we merge this one.

DarkLight1337 commented on vllm/engine/arg_utils.py:
```diff
@@ -622,12 +642,25 @@ def create_engine_config(self, ) -> EngineConfig:
+                self.image_processor = self.model
+            if self.no_image_processor:
+                if self.image_processor != self.model:
+                    raise ValueError(
+                        'Do not specify `image_processor` when it is disabled '
+                        'by `--no-image-processor`.')
```
I guess that would make it easier to quickly turn on/off image processing. I'll downgrade the error to a warning then.

DarkLight1337 committed:
Improve CLI args

DarkLight1337 committed:
Rename `multi_modal_datas` parameter

DarkLight1337 commented on vllm/worker/model_runner.py:
```diff
@@ -386,9 +397,17 @@ def _prepare_model_input(
-                if seq_group_metadata.multi_modal_data:
-                    multi_modal_input_list.append(
-                        seq_group_metadata.multi_modal_data.data)
+                mm_data = seq_group_metadata.multi_modal_data
```
It's used in the line `mm_kwargs = self.multi_modal_input_processor(mm_data)`.

ywang96 replied:
Yeah - what I meant really was that we can simply do since `mm_data` is not used anywhere else in this function:
`mm_kwargs = self.multi_modal_input_processor(seq_group_metadata.multi_modal_data)`

DarkLight1337 replied:
The variable is also used in the `if` statement. I made an intermediate variable to avoid exceeding the line limit in `mm_kwargs = self.multi_modal_input_processor(seq_group_metadata.multi_modal_data)`.

DarkLight1337 committed:
Rename `input_processor` to be more explicit

DarkLight1337 committed:
Rename `multi_modal_data` to be more explicit

DarkLight1337 committed:
Remove patch for LLaVA-NeXT

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Apply multi-modal refactor to `CPUModelRunner`

DarkLight1337 committed:
Fix multi-modal handling in `EmbeddingModelRunner`

DarkLight1337 commented:
> Thank you for the great work and I left a few comments! I've also reviewed your other PR #4910, so let's get that merged first then come back to this.
> 
> P.S I'm also working on a GPT4V API server PR as a follow up once we merge this one.

Thanks for your time! I have updated the code according to your comments. During this process, I noticed that I didn't update the code for some model runners so I fixed them as well.

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Move dummy image data generation to model-agnostic file

DarkLight1337 committed:
Add multimodal docs

DarkLight1337 commented on vllm/multimodal/registry.py:
```diff
@@ -0,0 +1,117 @@
+
+class MultiModalRegistry:
+
+    DEFAULT_PLUGINS = (ImageFeaturePlugin(), ImagePixelPlugin())
```
I have added a new section to the vLLM documentation to explain this.

DarkLight1337 committed:
Improve documentation for LLM/engine

DarkLight1337 commented:
Before merging this PR, imo we should complete #4328 first as it simplifies the API of passing multi-modal data.

DarkLight1337 committed:
Direct readers to the `PromptInputs` class

DarkLight1337 committed:
Separate `_run_engine` from `_validate_and_add_requests`

DarkLight1337 committed:
Add flag for deprecating legacy API

DarkLight1337 committed:
Add tests for `deprecate_kwargs`

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Rename attribute to be less misleading

DarkLight1337 committed:
Renable using `'fork'` start method and improve speed by using `torch.multiprocessing` wrapper instead of stdlib `multiprocessing`

DarkLight1337 committed:
Simplify logic of casting request output

DarkLight1337 committed:
Improve code readability

DarkLight1337 committed:
Fix `multi_modal_data` being a required key

DarkLight1337 committed:
Fix index out of range error

DarkLight1337 committed:
Use a flag to control whether to check output types

DarkLight1337 committed:
Simplify flags

DarkLight1337 committed:
Move output validation to a more appropriate location

DarkLight1337 committed:
Add message to deprecation notice

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Remove unused parameter in `_validate_and_add_requests` and fix test

DarkLight1337 committed:
Simplify code

DarkLight1337 committed:
Move attribute assignment outside `_init_tokenizer`

DarkLight1337 committed:
Only emit warning once

DarkLight1337 committed:
Simplify assignment expression

DarkLight1337 committed:
Place special case at the start

ywang96 committed:
move API reference to under developer doc

DarkLight1337 committed:
Fix links in docs

DarkLight1337 committed:
Remove unnecessary code to avoid repeated warning

DarkLight1337 committed:
Merge branch 'llm-inputs' into mm-data-processor

DarkLight1337 committed:
Update docs

DarkLight1337 committed:
Use intersphinx and avoid long default values

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 commented:
I have merged #4328 into this PR in advance, so you might see additional diffs. These diffs should disappear once #4328 is merged into `main`.

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Fix bad merge

DarkLight1337 committed:
Reinstate whitespace

DarkLight1337 committed:
Do not support multiple multimodal data in legacy API

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Fix bad config dict

DarkLight1337 commented:
Note that I have split out some of the changes in `conftest.py` to #5118.

DarkLight1337 committed:
Fix tests

DarkLight1337 committed:
Apply formatter

DarkLight1337 committed:
Remove `multi_modal_data` support in legacy API

DarkLight1337 commented:
Note that I have removed `multi_modal_data` from the legacy LLM API since this PR will break existing code that uses it anyway (due to the new method of constructing `MultiModalData`).

ywang96 commented on vllm/multimodal/base.py:
```diff
@@ -0,0 +1,125 @@
+    pass
+
+
+D = TypeVar("D", bound=MultiModalData)
+N = TypeVar("N", bound=Type["nn.Module"])
```
Why do we need to define these two? Are these defined for line length purposes?

DarkLight1337 replied:
They provide the generic type parameters used in `MultiModalPlugin` and `MultiModalRegistry`.

ywang96 commented on vllm/multimodal/image.py:
```diff
@@ -0,0 +1,120 @@
+logger = init_logger(__name__)
+
+
+def _get_seq_data(seq_len: int, vlm_config: VisionLanguageConfig):
+    token_ids = [vlm_config.image_token_id] * vlm_config.image_feature_size
+    token_ids += [0] * (seq_len - vlm_config.image_feature_size)
+
+    return SequenceData(token_ids)
```
Let's add a `NOTE` that this currently assumes image data is simply concatenated with prompt.

ywang96 commented on vllm/multimodal/image.py:
```diff
@@ -0,0 +1,120 @@
+logger = init_logger(__name__)
+
+
+def _get_seq_data(seq_len: int, vlm_config: VisionLanguageConfig):
```
Missing return type hint (same with a few other functions in this file)

ywang96 replied:
Is this only for dummy data generation? If so let's rename it
```suggestion
def _get_dummy_seq_data(seq_len: int, vlm_config: VisionLanguageConfig):
```

ywang96 commented on vllm/multimodal/image.py:
```diff
@@ -0,0 +1,120 @@
+    return SequenceData(token_ids)
+
+
+def _get_values(vlm_config: VisionLanguageConfig):
```
```suggestion
def _get_dummy_values(vlm_config: VisionLanguageConfig):
```

ywang96 commented on vllm/multimodal/image.py:
```diff
@@ -0,0 +1,120 @@
+                                                       vlm_config)
+        if image_processor is None:
+            image_arr = np.array(image, copy=True)
+            pixel_values = torch.as_tensor(image_arr) \
+                .view(1, image.height, image.width, -1) \
+                .permute((0, 3, 1, 2)) \
+                .to(model_config.dtype)
```
We should add a `NOTE` reference on this why this is the default way to get pixel values of an `PIL.Image.Image`

ywang96 commented on vllm/multimodal/registry.py:
```diff
@@ -0,0 +1,156 @@
+                                 vlm_config=vlm_config)
+
+
+MM_REGISTRY = MultiModalRegistry()
```
IMO the name `MM_REGISTRY` is slightly confusing and informal since we will be using this a lot.

What's the benefit of this over making all functions of `MultiModalRegistry` static methods and directly do something like
`@MultiModalRegistry.register_image_feature_input()`

DarkLight1337 replied:
Using a constant makes it more obvious that it's a global instance. I frequently encounter this while dealing with 3D vision models (e.g. [this repo](https://github.com/search?q=repo%3APointcept%2FPointcept%20register&type=code))

DarkLight1337 replied:
If `MM_REGISTRY` is too vague, we can use the full name `MULTIMODAL_REGISTRY`. Wdyt?

ywang96 replied:
I see - yea this makes sense then

ywang96 reviewed (commented):
@DarkLight1337 Left another round of reviews - PTAL, thanks!

DarkLight1337 committed:
Add NOTE and TODO

DarkLight1337 committed:
Add missing type annotations

DarkLight1337 committed:
Rename functions

DarkLight1337 committed:
Add NOTE

DarkLight1337 committed:
Fix multimodal inputs being on wrong device

DarkLight1337 committed:
Rename `MM_REGISTRY` to be more explicit

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 reviewed (commented):
Forgot to remove vllm/worker/model_runner.py#L531-533, but I turned off my PC already. Can you do it for me?

ywang96 committed:
fix upstream merge

ywang96 commented on examples/llava_example.py:
```diff
@@ -23,13 +24,11 @@ def run_llava_pixel_values():
+    image = Image.open("images/stop_sign.jpg")
 
     outputs = llm.generate({
-        "prompt":
-        prompt,
-        "multi_modal_data":
-        MultiModalData(type=MultiModalData.Type.IMAGE, data=image),
+        "prompt": prompt,
+        "multi_modal_data": ImagePixelData(image),
     })
```
Should we add an example here that if `--disable-image-processor` is specified, then we need to actually pass in pixel values (i.e, `images/stop_sign_pixel_values.pt`)?

DarkLight1337 replied:
Sure!

ywang96 commented on examples/llava_example.py:
```diff
@@ -23,13 +24,11 @@ def run_llava_pixel_values():
         "\nUSER: What is the content of this image?\nASSISTANT:")
 
     # This should be provided by another online or offline component.
```
```suggestion
```

ywang96 reviewed (commented):
Some final nits - once I test out the example again we should be good to go!

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Enable passing tensor directly as image

DarkLight1337 committed:
Add pillow to intersphinx and fix quote format

DarkLight1337 committed:
Fix mock imports

DarkLight1337 commented:
I have enabled tensors to be passed to `ImagePixelData` to skip preprocessing. However, it seems that the pixel tensor has slightly different preprocessing than HuggingFace's `LLaVAProcessor`, causing the test to fail.

DarkLight1337 committed:
Trigger pipeline

Trigger pipeline

Trigger pipeline

DarkLight1337 committed:
Automatically convert dtype

DarkLight1337 committed:
Comment out failing test for now

DarkLight1337 committed:
Fix blank pages in docs

DarkLight1337 committed:
Use the module name, not package name

DarkLight1337 committed:
Trigger pipeline

DarkLight1337 committed:
Trigger pipeline 2

ywang96 commented:
Per offline discussion - waiting for #5118 to be merged first.

DarkLight1337 committed:
Fix formatting [skip ci]

ywang96 commented:
@DarkLight1337 Could you resolve the merge conflicts? Once that's done I think this PR is ready to merge.

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

DarkLight1337 committed:
Merge branch 'upstream' into mm-data-processor

ywang96 commented on vllm/model_executor/models/llava.py:
```diff
@@ -82,6 +84,9 @@ class LlavaImageFeatureInputs(TypedDict):
 LlavaImageInputs = Union[LlavaImagePixelInputs, LlavaImageFeatureInputs]
 
 
+@MULTIMODAL_REGISTRY.register_image_feature_input()
+@MULTIMODAL_REGISTRY.register_image_pixel_input()
+@MULTIMODAL_REGISTRY.register_dummy_data(get_dummy_image_data)
```
Note for a later PR: we should add update doc regarding this additional registry step for adding new multi-modal language models

DarkLight1337 replied:
Will be added in #5276

ywang96 reviewed (approved):
I did a final pass and left a note, but everything LGTM! Thank you for the hard work on this. @DarkLight1337

nittaya3 commented:
Hello. I used my new profile yesterday. to add you guys And I want to see new people too. But I'm in France with you all the same.
